Draft (#1)
The-Marcy-Lab-School-Assignments/swe-sr-5-1-ajenec: README.md

Imagine you are teaching a friend about OOP. They mainly want to understand what is Encapsulation. Write a brief lesson on Encapsulation that includes the following:

- What is encapsulation?
- What major goal does this help to achieve in software engineering?
- Give an example (in code) of encapsulation.
- An explanation of how the code example demonstrates encapsulation

### Response 1

> In **object-oriented programming (OOP)**, `encapsulation` is all about bundling variables and the methods that operate on the data into a single, easy-to-use unit. Think of it like a medicine capsule. The capsule has everything you need to feel better, but you can't see what's inside. All you know is the label on the outside, which tells you what it is. Just like with `encapsulation`, data, and methods are grouped together in a private unit. It conceals the inner workings and exposes only what is necessary.

Author: **[Praise]** Such a strong start!

The major goal that encapsulation helps achieve in software engineering is to hide and protect data from being directly accessed or modified outside of the object.

- **Data Hiding**: It ensures that the object's internal data is protected and can only be accessed or modified through defined methods.

> - **Privacy**: It hides how the object is modified and allows the user of the object to focus only on what the object does.

Author: **[Clarity]** This feels redundant of both "data hiding" and "control over data"

- **Easy to Maintain**: Makes it easier to change the internal implementation without affecting the users of the object.
- **Control Over Data**: Allows the class to enforce rules about how data is accessed or modified.

```js
// This is the object
const makeBucketList = {
  toDo: [],

  // This function is the "method", which is stored inside the object
  addToList(newToDo) {

>     // `this` refers to the "owner" of the method, which is the object it is invoked

Author: ```suggestion
    // `this` refers to the "owner" of the method, which is the object it is invoked on
```

    this.toDo.push(newToDo);
  },
  // This method allows you to see the data without directly accessing it
  printList() {
    console.log(this.toDo);
  },
};
```

> **Data Hiding**: The `toDo` property is _encapsulated_ inside of the `makeBucketList` object. This property is not directly accessible from outside the object. You need to use the methods (`addToList` and `printList`).

Author: **[Accuracy]** This isn't true. The `toDo` property looks like it is a property of the object, therefore making it accessible from the outside. In order to hide this property, you'd need either make a factory function and use closure to hide the todo array or use a class with a private property.

> **Access Control**: The `addToList` method is responsible for modifying the `toDo` array, and the `printList` method is used access and display the data. By controlling the access to the `toDo` array through these methods, it ensures that the data in the object cannot be modified outside the object easily. Which is the key of **encapsulation**.

Author: **[Praise]** This is excellent!

```suggestion
**Access Control**: The `addToList` method is responsible for modifying the `toDo` array, and the `printList` method is used access and display the data. By controlling the access to the `toDo` array through these methods, it ensures that the data in the object cannot be modified outside the object easily, which is the key of **encapsulation**.
```

## Prompt 2

The following `friendsManager` object is an example of an interface that is **NOT** consistent and predictable:

```js
const friendsManager = {
  friends: [],
  addFriend(newFriend) {
    if (typeof newFriend !== "string") return;
    this.friends.push(newFriend);
  },
};

friendsManager.addFriend("daniel");
friendsManager.addFriend(true);
friendsManager.friends.push("emmaneul");
friendsManager.friends.push(42);
```

Explain how the code is not consistent or predictable, then provide an example in code that uses closure to make it more consistent and predictable.

### Response 2

> The `friendsManager` function has some issues that prevents it from being consistent and predictable. The `friends` array is publicly accessible, meaning anyone can change it using `friendsManager.friends.push("friend")`. Since the array is publicly exposed, the user can bypass `addFriend()` method and push any data type (including non-string, boolean, etc.) values to the array which leads to inconsistent behavior of the output. We want to set up the `friends` array so that it can be modified only using the `addFriend()` method and not bypass any restrictions we specify in our method. Also `friendsManager` is a plain object and not a class-based or factory-based structure that allows to create multiple instances.

Author: **[Writing]** This is excellent writing! Adding a few suggested changes to improve things slightly.

```suggestion
The `friendsManager` function has some issues that prevent it from being consistent and predictable. The `friends` array is publicly accessible, meaning anyone can change it using `friendsManager.friends.push("friend")`. Since the array is publicly exposed, the user can bypass the `addFriend()` method and push values of any data type (including non-string, boolean, etc.) to the array which leads to inconsistent behavior. We want to set up the `friends` array so that it can be modified only using the `addFriend()` method and not bypass any restrictions we specify in our method. Also, `friendsManager` is a plain object and not a class-based or factory-based structure that would allow us to create multiple instances.
```

> Here is an example of the revised code above that is consistent with OOP guidelines and is more predictable:

Author: **[Praise]** Wonderful example y'all!

```js
class FriendsManager {
  // creates a private array where we can store all the friends of the individual instances
  #friends = [];

  addFriend(newFriend) {
    if (typeof newFriend !== "string") {
      console.log("Only strings allowed bro...");
      return;
    }
    this.#friends.push(newFriend);
  }

  getFriends() {
    // returns a shallow copy of the array to prevent the direct mutation
    return [...this.#friends];
  }
}

// creates a new instance
const danny = new FriendsManager();

// adds a new friend
danny.addFriends("Josh");

// does not add new friend because of the validation
danny.addFriends(213);

// returns ["Josh"]
console.log(danny.getFriends());

// throws a syntax error because cannot access the `friends` array directly
console.log(danny.#friends);
```

## Prompt 3

With OOP in JavaScript, it's possible to use factory functions to achieve encapsulation and re-use them to make objects that look alike. However, factory functions have drawbacks and we often use classes instead.

How would you explain to a budding developer what the drawbacks of using factory functions are and why it is better to use classes instead?

### Response 3

> A `factorty function` is functions in JavaScript that can return objects. Here's an example:

Author: **[Typos]**

```suggestion
A `factory function` is a function in JavaScript that can return an object. Here's an example:
```

```js
const favFoodsList = (...initialList) => {
  // Creates a new array 'list' that copies the passed initial foods
  const list = [...initialList];

  // Returns an object with two methods:
  return {
    // This method returns the current list
    getList() {
      return list;
    },

    // This method allows you to add to the list
    addToList(newFav) {
      list.push(newFav);
    },
  };
};

// Calling the factory function to create a new object
const myList = favFoodsList("french toast", "pancakes");

myList.addToList("fruit"); // Adds to current list: 'myList'

> console.log(myList.getList()); // This logs: [ 'french toast', 'pancakes', 'fruit' ]

Author: **[Praise]** Nice Example!

```

Every time the **factory function** is invoked, a brand new object is made with the same bahavior. Each object holds a seperate copy of the same methods. This increases memory usage because methods are stored independently for each object instead of sharing the method.

A `class` is a blueprint of something with variables and functions, used to create an object and execute its behavior. They are more structured and designed for managing objects.

```js
class GreetMe {
  greeting() {
    console.log("Hey Dude!");
  }
}

const buddy = new GreetMe();

> buddy.greeting(); // Output: Hey Dude!

Author: **[Example]** If this example were a `class` version of the FavFoodsList factory function, then it would more clearly demonstrate the difference between the two.

```

> With **Classes** you can easily create subclasses that inherit properties and methods from the parent class. This allows you to reuse the code and keep things organized. When you create an object using a `class`, it uses the same set of methods making the code easier to maintain.

Author: **[Structure]** Don't forget to mention the memory usage! This was the main downside of factory functions that you mentioned above. If you forget to mention it here, the reader will forget about that critical reason as well.

```suggestion
With **Classes** you can easily create subclasses that inherit properties and methods from the parent class. This allows you to reuse the code and keep things organized. When you create an object using a `class`, it uses the same set of methods making the code easier to maintain and take up less memory.
```

## Prompt 4

Do some research on the history of when / how classes were introduced into JavaScript and share your findings. Your response should include:

- What version of JavaScript were classes introduced in and when did it come out?
- Why were classes introduced into JavaScript?

### Response 4

> `Classes` were introduced in ES6 version of JavaScript and it came out in June 2015. Along with classes this version came out with a few more syntax features.

Author: **[Formatting]** Here, you use backticks to wrap around `Classes` which isn't really appropriate. Reserve using backticks for when you are referencing keywords in code or code expressions. If you just want to highlight a key term, use bolding with double asterisks.

1.  Class syntax (`class`, `constructor`, `extends`, `super`)
2.  Arrow functions (`=>`)
3.  `let` and `const` for block-scoped variables
4.  Template literals (backticks `Hello ${name}`)
5.  Default parameters
6.  Destructuring assignment
7.  Promises

> 8.  Modules (`import`/`export`)

Author: **[Clarity]** While nice to know, these other features aren't necessary here.

    this.name = name;
    this.age = age;
  }

  greet() {
    console.log(`Hello, my name is ${this.name}.`);
  }
}
```

This is much cleaner syntax and resembles a traditional OOP languages.

`Classes` allow defining objects and dealing with inheritance much easier.
Here is an example of prototype-based inheritance before ES6:

```js
function Employee(name, age, jobTitle) {
  Person.call(this, name, age); // Call parent constructor
  this.jobTitle = jobTitle;
}

Employee.prototype = Object.create(Person.prototype);
Employee.prototype.constructor = Employee;

Employee.prototype.work = function () {
  console.log(`${this.name} is working as a ${this.jobTitle}.`);
};
```

Here is a better version using `Classes` with (`extends` & `super`)

```js
class Employee extends Person {
  constructor(name, age, jobTitle) {
    super(name, age); // Calls the parent class constructor
    this.jobTitle = jobTitle;
  }

  work() {
    console.log(`${this.name} is working as a ${this.jobTitle}.`);
  }
}
const emp1 = new Employee("Bob", 25, "Software Engineer");
emp1.greet(); // Inherited from Person class
emp1.work(); // Output: Bob is working as a Software Engineer.
```

> This syntax is much easier to read and handle inheritance.

Author: **[Praise]** This is wonderful! I love how you compare and contrast the two approaches a few times to drive home how much cleaner `class` syntax is.

};

const alice = new Person("Alice", 30);
console.log(alice.greet());
```

Provide one point that advocates for the use of this syntax and then provide a counter-argument for the use of classes instead.

### Response 5

**Point for Constructure Functions and Prototype:**

Using `Constructure Functions` and the `Prototype Chain` allows for more flexibility. You dont have to use the class keyword, which adds more structure. **Constructure Functions** allow you to create objects and define methods in a straightforward way. The **Prototype Chain** makes inheritance more transparent and customizable, giving you more control over how objects inherit behaviors.

```js
// Constructor function to create a 'Person' object
function Person(name, age) {
  this.name = name;
  this.age = age;
}

// Adding a method to the prototype chain
Person.prototype.greet = () => {
  console.log(`Hello, my name is ${this.name}`);
};

// Creating an instance of Person
const ann = new Person("Ann", 24);
// Calling the greet method (using the prototype chain)
ann.greet(); // Output: 'Hello, my name is undefined'
```

This shows the potential pitfall when working with prototype chain.

> - The `Person.prototype.greet` doesnt bind its own `this`, which means it refers to the surrounding context instead of the instance `ann` leading to `this.name` being undefined.

Author: **[Clarity]** I'm confused about this, I can't tell if using this is arguing that using the prototype chain is better or worse. The writing is clear enough because it seems like you're saying that by NOT using the `class` syntax, it is more structured but then the example demonstrates a pitfall of using the prototype chain rather than a benefit?
